src/drag_icon.c

```diff
@@ -0,0 +1,80 @@
+// SPDX-License-Identifier: GPL-2.0-only
+#include <assert.h>
+#include <wlr/types/wlr_data_device.h>
+#include <wlr/types/wlr_scene.h>
+#include <wlr/util/log.h>
+#include "drag_icon.h"
+#include "labwc.h"  /* for struct seat */
+
+static void
+on_icon_map(struct wl_listener *listener, void *data)
+{
+	struct drag_icon *self = wl_container_of(listener, self, events.map);
+	struct wlr_drag_icon *icon = data;
+	icon->data = wlr_scene_subsurface_tree_create(self->tree, icon->surface);
```

Comment:
Should check for `icon->data` already being set..Otherwise map -> unmap -> map feels a bit clunky (won't leak memory though because we are destroying the shared parent).